src/components/views/settings/ChangePassword.tsx

```diff
@@ -82,48 +82,64 @@ export default class ChangePassword extends React.Component<IProps, IState> {
         };
     }
 
-    private onChangePassword(oldPassword: string, newPassword: string): void {
+    private async onChangePassword(oldPassword: string, newPassword: string): Promise<void> {
         const cli = MatrixClientPeg.get();
 
-        if (!this.props.confirm) {
-            this.changePassword(cli, oldPassword, newPassword);
-            return;
-        }
+        const serverSupportsControlOfDevicesLogout: boolean = await cli.doesServerSupportLogoutDevices();
 
-        Modal.createTrackedDialog('Change Password', '', QuestionDialog, {
-            title: _t("Warning!"),
-            description:
-                <div>
-                    { _t(
-                        'Changing password will currently reset any end-to-end encryption keys on all sessions, ' +
-                        'making encrypted chat history unreadable, unless you first export your room keys ' +
-                        'and re-import them afterwards. ' +
-                        'In future this will be improved.',
-                    ) }
-                    { ' ' }
-                    <a href="https://github.com/vector-im/element-web/issues/2671" target="_blank" rel="noreferrer noopener">
-                        https://github.com/vector-im/element-web/issues/2671
-                    </a>
-                </div>,
-            button: _t("Continue"),
-            extraButtons: [
-                <button
-                    key="exportRoomKeys"
-                    className="mx_Dialog_primary"
-                    onClick={this.onExportE2eKeysClicked}
-                >
-                    { _t('Export E2E room keys') }
-                </button>,
-            ],
-            onFinished: (confirmed) => {
-                if (confirmed) {
-                    this.changePassword(cli, oldPassword, newPassword);
-                }
-            },
-        });
+        if (serverSupportsControlOfDevicesLogout) {
+            // don't log user out of all devices
+            this.changePassword(cli, oldPassword, newPassword, /* logoutDevices = */ false);
+        } else {
```

Comment:
This block (won't let me select to the other end of the else) is a pain to read, Modal.createTrackedDialog has an async signature which you can use to DRY this area.

e.g

```javascript
const logoutDevices = serverSupportsControlOfDevicesLogout ? false : undefined;

if (!serverSupportsControlOfDevicesLogout && this.props.confirm) {
    const { finished } = Modal.createTrackedDialog<[boolean]>(...);
    const [confirmed] = await finished;
    if (!confirmed) return;
}

this.changePassword(cli, oldPassword, newPassword, logoutDevices);
```

or similar

Reply:
Much nicer. Thanks!